Use redux

datalater opened the pull request:
4주차 1번 과제 제출합니다.

## 진행상황

### 2020.02.16(화)

아직 완성되지 않았습니다만 일단 작성된 부분까지 제출합니다.

* App 컴포넌트에 있는 상태 관리 코드를 리덕스로 바꿀 것이다.
* App 컴포넌트에 있는 상태 관리 코드를 분리한 후 리덕스로 바꿀 것이다.
* 따라서 App 컴포넌트 상태 관리 코드를 분리하는 리팩토링 과정이 필요하다.
* 그런데 리팩토링 하기 전에 리팩토링이 잘 되는지 검사해줄 테스트 코드가 없다.
* 따라서 App 컴포넌트의 상태 관리 코드에 대한 테스트를 작성한다.

### 2020.02.17(수)

* reducer 테스트 코드를 작성한다.
* reducer를 구현한다.
* App 컴포넌트에 있던 상태 관리 코드를 분리할 것이다.
* reducer를 만들어서 상태 변경 로직을 모두 reducer 파일로 분리한다.
* action creator 메소드를 actions 파일로 분리한다.
* Page에 집중된 책임을 분산하기 위해 ListContainer 테스트 코드를 작성한다.
* ListContainer를 구현한다.
* ListContainer 생성에 따라 App 및 Page 컴포넌트를 리팩토링한다.
* Page에 집중된 책임을 분산하기 위해 InputContainer 테스트 코드를 작성한다.
* InputContainer를 구현한다.
* InputContainer 생성에 따라 App 및 Page 컴포넌트를 리팩토링한다.

### 2020.02.19(금)

* 테스트가 항상 일정한 상태로 시작되도록 mock 함수를 beforeEach 블록문에 넣는다.
* 여러 개의 if 문을 object literal로 리팩토링한다.
* 테스트 코드에서 context를 대칭적으로 구성하여 작성한다.

### 2020.02.21(일)

* 중복 테스트를 제거한다.

## TODO

## 궁금한 점

## 요청 받은 코드 리뷰 목록

## 메타인지

### 내가 모르는 것은 무엇인가

* useDispatch는 mockImplementaion으로 어떻게 구현하는가?

### 내가 알게 된 것은 무엇인가

### 어떻게 하면 더 나을까
 
## 주간회고

datalater committed:
App에 대한 updtate task title 테스트를 추가하라

redux를 구현하려면 redux에 대한 테스트가 필요하다.
그런데 바로 redux에 대한 테스트를 작성하는 것보다 App 컴포넌트에서 상태
관리 코드를 분리한 이후에 redux로 넘어가는 것이 더 간단하다.
하지만 App 컴포넌트 테스트 코드는 상태 관리 코드에 대한 테스트 코드가
없으므로 이에 대한 테스트 코드 먼저 작성한다.

datalater committed:
App에서 update task title 관련 코드를 리팩토링 하라

updtae task title 관련 코드를 함수로 분리한다.

datalater committed:
App에 대한 add task 테스트를 추가하라

datalater committed:
App에서 add task 관련 코드를 리팩토링 하라

add task 관련 상태 관리 코드를 함수로 분리한다.

datalater committed:
App에 대한 delete task 테스트를 추가하라

datalater committed:
App에서 delete task 관련 코드를 리팩토링 하라

delete task 관련 상태 관리 코드를 함수로 분리한다.

datalater committed:
App 테스트를 리팩토링 하라

중복되는 render 함수 호출 코드를 함수로 분리한다.

datalater committed:
App 테스트 코드에서 불필요한 주석을 제거하라

datalater committed:
reducer의 updateTaskTitle 테스트 코드를 추가하라

Kihyun92 reviewed (commented):
커밋 메시지를 잘 작성해주시는 것 같아요! 👍🏼 

강의를 보고 따라하는것도 도움이 되지만, 직접 해보는건 훨씬 도움이 되죠 ㅎㅎ

배웠던 것을 떠올려 보면서 과제를 진행해보시면 좋을 것 같습니다 :)

datalater commented:
네 강의 보고 따라 치고 있지 않고요 ㅎㅎ 복기하면서 하고 있습니다

datalater committed:
reducer의 updateTaskTitle을 구현한다

전달받은 action이 updateTaskTitle이면 taskTitle을 payload에서 받아
상태를 갱신한다.

datalater committed:
reducer의 addTask 테스트 코드를 추가하라

datalater committed:
reducer의 addTask를 구현한다

datalater committed:
reducer의 deleteTask 테스트 코드를 추가하라

datalater committed:
reducer의 deleteTask를 구현한다

datalater committed:
reducer 테스트 코드를 리팩터링 하라

공백줄을 추가하여 가독성을 늘리고 updateTaskTitle의 테스트 이름을 더
간단하게 변경한다.

datalater committed:
reducer addTask에서 taskTitle이 없는 경우에 대해 테스트를 작성하라

datalater committed:
reducer에서 taskTitle이 없는 경우에 대해 addTask를 구현하라

datalater committed:
App 컴포넌트에 존재하는 상태 관리 코드를 redux로 변경하라

- store를 적용하여 App 컴포넌트에서 useSelector로 필요한 state를
  가져온다.
- setState 대신 useDispatch를 사용하여 store에 action을 전달한다.
- App 컴포넌트에 있는 상태 변경 로직은 reducer에 있으므로 제거한다.
- App 컴포넌트에 있는 초기 상태값은 은 reducer로 옮긴다.
- dispatch에서 사용할 action creator 메소드를 actions에서 불러온다.
- store를 사용할 수 있도록 App 컴포넌트를 Provider로 감싼다.

datalater committed:
reducer에서 등록되지 않은 action에 대해 테스트하라

datalater committed:
불필요한 주석을 삭제하라

datalater committed:
reducer에서 초기 상태값을 사용하는 테스트를 작성하라

datalater commented:
# 질문

![image](https://user-images.githubusercontent.com/8105528/108165468-f3a7d500-7135-11eb-97f3-a4e4cc67ce28.png)

위 커버리지를 보면 App 컴포넌트에서 핸들러 함수가 테스트 되지 않은 부분이 있습니다. 리덕스 container 컴포넌트, presentational 컴포넌트로 리팩토링하기 전에 현재 코드에서 App 테스트 커버리지를 100%로 채워보고 싶습니다. 

그런데 useSelector는 가짜로 구현했지만 useDispatch를 가짜로 구현하지 않아서 dispatch가 함수가 아니라는 오류가 뜹니다.

useDispatch를 useSelector처럼 mockImplementation으로 구현해보고 싶은데 이렇게 되면 store도 가짜로 또 구현해야 하는 게 아닌가 싶어요. 어떻게 접근해보면 될까요?

datalater committed:
App 컴포넌트에서 updateTaskTitle을 테스트하라

datalater commented:
dispatch가 함수가 아니라는 오류가 발생하는 상황을 다시 재현하기 위해 5ba6dca 커밋을 추가했습니다.

datalater commented:
# 질문

![image](https://user-images.githubusercontent.com/8105528/108166704-ea1f6c80-7137-11eb-864b-76195137b1b3.png)


위 커버리지를 보면 App 컴포넌트에서 핸들러 함수가 테스트 되지 않은 부분이 있습니다. 리덕스 container 컴포넌트, presentational 컴포넌트로 리팩토링하기 전에 현재 코드에서 App 테스트 커버리지를 100%로 채워보고 싶습니다. 

그런데 useSelector는 가짜로 구현했지만 useDispatch를 가짜로 구현하지 않아서 dispatch가 함수가 아니라는 오류가 뜹니다.

![image](https://user-images.githubusercontent.com/8105528/108166801-0b805880-7138-11eb-802c-6ff126abf084.png)

useDispatch를 useSelector처럼 mockImplementation으로 구현해보고 싶은데 이렇게 되면 store도 가짜로 또 구현해야 하는 게 아닌가 싶어요. 어떻게 접근해보면 될까요?

datalater committed:
redux 사용에 따라 App 테스트 코드를 변경하라

datalater committed:
ListContainer의 delete task 테스트를 작성하라

datalater committed:
ListContainer의 delete task를 구현하라

datalater committed:
ListContainer 생성에 따라 App 및 Page 컴포넌트를 리팩터링 하라

ListContainer가 분담하는 책임을 App과 Page 컴포넌트에서 제거한다.

datalater committed:
InputContainer 테스트를 작성하라

datalater committed:
InputContainer를 구현하라

datalater committed:
Page 컴포넌트 로직을 App 컴포넌트가 대신하도록 리팩토링하라

datalater commented:
테스트 커버리지 결과를 첨부합니다.

![image](https://user-images.githubusercontent.com/8105528/108178574-edbaef80-7147-11eb-99ab-16f18a6c5481.png)

Kihyun92 commented on src/App.test.jsx:
```diff
@@ -2,15 +2,34 @@ import React from 'react';
+  }
 
-  expect(getByText(/추가/)).not.toBeNull();
+  it('shows "추가" text', () => {
```
실제론 버튼이니까 `shows "추가" button`은 어떤가요?

datalater replied:
그렇네요 ㅎㅎ 반영하겠습니다!

datalater replied:
2a72a76

Kihyun92 commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,65 @@
+describe('InputContainer', () => {
+  const dispatch = jest.fn();
+
+  useDispatch.mockImplementation(() => dispatch);
+
+  useSelector.mockImplementation((selector) => selector({
+    newId: 100,
+    taskTitle: '',
+  }));
```
`beforeEach`함수를 사용해서 테스트가 실행될 때마다 항상 일정한 상태에서 시작할 수 있도록 해주는게 좋을 것 같아요.
그러지 않으면 테스트가 다른 테스트들에 영향을 미칠 수 있고, 그러면 테스트가 올바르게 동작하지 않거나 유지보수하기 힘들어질 거에요.

```
beforeEach(() => {
  useDispatch.mockImplementation(() => dispatch);

  useSelector.mockImplementation((selector) => selector({
    taskTitle: 'New Title',
  }));
});

```

datalater replied:
아하 감사합니다!!

datalater replied:
df7a985

Kihyun92 commented on src/reducer.js:
```diff
@@ -0,0 +1,40 @@
+};
+
+export default function reducer(state = initialState, action) {
+  if (action.type === 'updateTaskTitle') {
+    return {
+      ...state,
+      taskTitle: action.payload.taskTitle,
+    };
+  }
+
+  if (action.type === 'addTask') {
+    const { newId, tasks, taskTitle } = state;
+
+    if (taskTitle === '') {
+      return state;
+    }
+
+    return {
+      ...state,
+      newId: newId + 1,
+      taskTitle: '',
+      tasks: [...tasks, { id: newId, title: taskTitle }],
+    };
+  }
+
+  if (action.type === 'deleteTask') {
+    const { tasks } = state;
+
+    return {
+      ...state,
+      tasks: tasks.filter((task) => task.id !== action.payload.id),
+    };
+  }
```
`if`나 `switch`를 사용하지 않고 해결할 수 있을까요?

datalater replied:
object literal을 사용하도록 변경해보겠습니다!

datalater replied:
22d7fc2

Kihyun92 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,94 @@
+      expect(state.taskTitle).toBe('new-title');
+    });
+
+    context('without previous state offered', () => {
```
이 경우엔 `context`보단 `it`으로 처리하는게 맞을 것 같습니다. 혹은 위쪽에 테스트를 `context`로 감싸주는게 좋을 것 같아요.

다른 테스트들도 확인해주세요~

`with <-> without`
`when log in <-> when log out`
처럼 같은 맥락에서 별개의 사례일 때 `context`를 사용합니다.

context 사용법은 아래 문서와 아샬님의 문서를 참고해주세요.

- https://www.betterspecs.org/
- https://github.com/ahastudio/til/blob/main/blog/2018/12-08-given-when-then.md

datalater replied:
아하.. 대칭적일 때 사용하는 것이라 생각하면 되겠군요. 감사합니다! 반영할게요.

datalater replied:
c386d31

Kihyun92 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,94 @@
+    });
+  });
+
+  describe('unregisteredAction', () => {
```
이 부분은 보다 상위에서 context로 처리해주는건 어떨까요? 

``` jest
context('when has unregistered action type', () => { ... })

context('when has action type', () => {
  // updateTaskTitle, addTask 등 처리
})
```

datalater replied:
이해되었습니다. 감사합니다. 반영할게요!

datalater replied:
그러면 c386d31 커밋처럼

```jsx
describe('reducer', () => {
  context('with unregistered action type', () => {
  })

  context('with registered action type', () => {
    describe('updateTaskTitle', () => {
      context('without previous state', () => {
        it('changes task title using initalState', () => {});
      });

      context('with previous state', () => {
        it('changes task title', () => {});
      });
    });
  })

})
```

위와 같이 될 것 같은데, `describe-context-describe-context-it` 이런 식으로 `context`와 `describe`의 순서가 뒤바뀌는 식으로 사용해도 괜찮은가요?

위와 같은 형태로 진행한 테스트 실행 결과를 첨부합니다.

![image](https://user-images.githubusercontent.com/8105528/108499617-edf6ee80-72f1-11eb-9c96-cc407fc90726.png)

Kihyun92 replied:
개인적으로는 중간에 `without previous state offered ~~`를 제외하면 계층구조로 잘 나뉜 것 같다고 생각이 드는것 같긴해요. 😅

그런데 중요한건 내가 테스트를 통해 뭘 보여주고 싶은지 같아요.ㅎㅎ 본인이 보기에 더 명확하게 전달이 가능한 구조로 테스트를 구현해보는게 좋을 것 같습니다😊

datalater replied:
그러네요 `without previous state offered ~~` 이 부분을 보니 테스트가 중복되어 있네요 수정해서 올리겠습니다!

피드백 감사합니다!!

datalater committed:
테스트 이름을 변경하라라

추가라는 텍스트가 있는지 테스트하는 것보다 추가라는 버튼이 있는지
테스트하는 것이 더 목적에 부합하다.

datalater committed:
테스트가 항상 일정한 상태로 시작되도록 변경한다

mock 함수를 구현하는 내용을 beforeEach 블록 안으로 넣는다.

datalater committed:
여러 개의 if 문을 object literal로 리팩토링하라

datalater committed:
이전 상태가 주어졌을 때와 그렇지 않을 때로 테스트 context를 구분하라

Kihyun92 reviewed (approved):
좋습니다! TDD cycle에 조금 익숙해지셨나요? ㅎㅎ 

[테스트 구조 관련해서 커맨트](https://github.com/CodeSoom/react-week4-assignment-1/pull/48#discussion_r579273685) 남겼습니다. 확인해보시고 여유가 되시면 수정해보셔도 좋을 것 같아요 :)

datalater commented:
테스트 구조 수정했습니다. 놓치고 있던 포인트들 피드백 주셔서 정말 감사합니다!!

![image](https://user-images.githubusercontent.com/8105528/108613863-1f2e0680-7439-11eb-958e-ee144cd2d5b9.png)

datalater committed:
중복된 테스트를 제거하라
